File: project-{topic}-enhanced/app/services/user_service.py
from app import db
from app.models.user import User
from app.models.cart import Cart
from app.utils.errors import NotFoundError, ConflictError, BadRequestError
import logging

logger = logging.getLogger(__name__)

class UserService:
    @staticmethod
    def get_all_users():
        return User.query.all()

    @staticmethod
    def get_user_by_id(user_id):
        user = User.query.get(user_id)
        if not user:
            raise NotFoundError(f"User with id {user_id} not found.")
        return user

    @staticmethod
    def create_user(username, email, password, role='customer'):
        if User.query.filter_by(username=username).first():
            raise ConflictError("Username already exists.")
        if User.query.filter_by(email=email).first():
            raise ConflictError("Email already registered.")

        try:
            new_user = User(username=username, email=email, password=password, role=role)
            db.session.add(new_user)
            db.session.flush() # To get the new_user.id for cart creation
            
            # Create a cart for the new user
            new_cart = Cart(user_id=new_user.id)
            db.session.add(new_cart)
            
            db.session.commit()
            logger.info(f"User {new_user.username} and their cart created successfully.")
            return new_user
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error creating user {username}: {e}")
            raise BadRequestError(f"Failed to create user: {e}")

    @staticmethod
    def update_user(user_id, data):
        user = UserService.get_user_by_id(user_id)
        
        if 'username' in data and data['username'] != user.username and User.query.filter_by(username=data['username']).first():
            raise ConflictError("Username already taken.")
        if 'email' in data and data['email'] != user.email and User.query.filter_by(email=data['email']).first():
            raise ConflictError("Email already registered.")

        try:
            user.username = data.get('username', user.username)
            user.email = data.get('email', user.email)
            if 'password' in data:
                user.password_hash = user.set_password(data['password'])
            user.role = data.get('role', user.role)
            user.is_active = data.get('is_active', user.is_active)
            db.session.commit()
            logger.info(f"User {user_id} updated successfully.")
            return user
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error updating user {user_id}: {e}")
            raise BadRequestError(f"Failed to update user: {e}")

    @staticmethod
    def delete_user(user_id):
        user = UserService.get_user_by_id(user_id)
        try:
            db.session.delete(user)
            db.session.commit()
            logger.info(f"User {user_id} deleted successfully.")
        except Exception as e:
            db.session.rollback()
            logger.error(f"Error deleting user {user_id}: {e}")
            raise BadRequestError(f"Failed to delete user: {e}")

File: project-{topic}-enhanced/src/services/UserService.cpp
```cpp
#include "UserService.hpp"
#include "../logger/Logger.hpp"
#include "../models/User.hpp"
#include "../database/Database.hpp" // For SQLiteException

#include <stdexcept>
#include <string>
#include <vector>
#include <optional>

UserService::UserService(Database& db) : db(db) {}

// Creates a new user in the database.
int UserService::createUser(User& user) {
    // Basic validation
    if (user.getUsername().empty() || user.getPasswordHash().empty() || user.getEmail().empty()) {
        throw std::runtime_error("Username, password hash, and email cannot be empty.");
    }
    // More complex validation (e.g., username/email uniqueness, password strength)
    // should happen in AuthService before calling this.

    try {
        int newUserId = User::create(db, user);
        user.setId(newUserId); // Update the user object with the new ID
        Logger::info("UserService: Created user '{}'. ID: {}", user.getUsername(), newUserId);
        return newUserId;
    } catch (const SQLiteException& e) {
        Logger::error("UserService: SQLite error creating user: {}", e.what());
        throw std::runtime_error("Database error while creating user.");
    }
}

// Retrieves a user by their ID.
std::optional<User> UserService::findById(int userId) {
    try {
        std::optional<User> user = User::findById(db, userId);
        if (user) {
            Logger::debug("UserService: Retrieved user with ID {}.", userId);
        } else {
            Logger::debug("UserService: User with ID {} not found.", userId);
        }
        return user;
    } catch (const SQLiteException& e) {
        Logger::error("UserService: SQLite error retrieving user by ID {}: {}", userId, e.what());
        throw std::runtime_error("Database error while retrieving user.");
    }
}

// Retrieves a user by their username.
std::optional<User> UserService::findByUsername(const std::string& username) {
    try {
        std::optional<User> user = User::findByUsername(db, username);
        if (user) {
            Logger::debug("UserService: Retrieved user with username '{}'.", username);
        } else {
            Logger::debug("UserService: User with username '{}' not found.", username);
        }
        return user;
    } catch (const SQLiteException& e) {
        Logger::error("UserService: SQLite error retrieving user by username '{}': {}", username, e.what());
        throw std::runtime_error("Database error while retrieving user.");
    }
}

// Retrieves a user by their email.
std::optional<User> UserService::findByEmail(const std::string& email) {
    try {
        std::optional<User> user = User::findByEmail(db, email);
        if (user) {
            Logger::debug("UserService: Retrieved user with email '{}'.", email);
        } else {
            Logger::debug("UserService: User with email '{}' not found.", email);
        }
        return user;
    } catch (const SQLiteException& e) {
        Logger::error("UserService: SQLite error retrieving user by email '{}': {}", email, e.what());
        throw std::runtime_error("Database error while retrieving user.");
    }
}

// Retrieves all users from the database.
std::vector<User> UserService::getAllUsers() {
    try {
        std::vector<User> users = User::findAll(db);
        Logger::debug("UserService: Retrieved all {} users.", users.size());
        return users;
    } catch (const SQLiteException& e) {
        Logger::error("UserService: SQLite error retrieving all users: {}", e.what());
        throw std::runtime_error("Database error while retrieving all users.");
    }
}

// Updates an existing user in the database.
bool UserService::updateUser(int userId, User& updatedUser) {
    // Ensure the ID of the updatedUser object matches the ID passed
    updatedUser.setId(userId);

    // Basic validation (e.g., username not empty)
    if (updatedUser.getUsername().empty() || updatedUser.getEmail().empty()) {
        throw std::runtime_error("Username and email cannot be empty for update.");
    }

    try {
        bool success = User::update(db, updatedUser);
        if (success) {
            Logger::info("UserService: Updated user with ID {}.", userId);
        } else {
            Logger::warn("UserService: User with ID {} not found for update.", userId);
            throw std::runtime_error("User not found for update.");
        }
        return success;
    } catch (const SQLiteException& e) {
        Logger::error("UserService: SQLite error updating user {}: {}", userId, e.what());
        throw std::runtime_error("Database error while updating user.");
    }
}

// Deletes a user by their ID.
bool UserService::deleteUser(int userId) {
    try {
        bool success = User::remove(db, userId);
        if (success) {
            Logger::info("UserService: Deleted user with ID {}.", userId);
        } else {
            Logger::warn("UserService: User with ID {} not found for deletion.", userId);
            throw std::runtime_error("User not found for deletion.");
        }
        return success;
    } catch (const SQLiteException& e) {
        Logger::error("UserService: SQLite error deleting user {}: {}", userId, e.what());
        throw std::runtime_error("Database error while deleting user.");
    }
}
```